vllm/v1/attention/backends/mamba2_attn.py

```diff
@@ -172,4 +135,95 @@ def build(self,
             chunk_offsets_p=chunk_offsets_p,
             state_indices_tensor=state_indices_tensor,
         )
-        return attn_metadata
+        return update_metadata(
+            attn_metadata) if num_prefills > 0 else attn_metadata
+
+
+def _query_start_loc_to_chunk_indices_offsets(query_start_loc: torch.Tensor,
+                                              chunk_size: int,
+                                              total_seqlens: int):
+
+    cu_seqlens = query_start_loc[1:]  # remove prepended 0
+
+    # outputs will have length expansion of chunks that do not divide
+    # chunk_size
+    N = math.ceil(total_seqlens / chunk_size) + (cu_seqlens[:-1] % chunk_size
+                                                 > 0).sum()
+    chunk_indices = torch.arange(N,
+                                 dtype=torch.int,
+                                 device=query_start_loc.device)
+    chunk_offsets = torch.zeros((N, ),
+                                dtype=torch.int,
+                                device=query_start_loc.device)
+
+    p = 0  # num of insertions
+    for s, e in zip(cu_seqlens[:-1], cu_seqlens[1:]):
+
+        # if does not divide chunk_size, then there is one chunk insertion
+        p += (s % chunk_size > 0)
+
+        # get the dimensions
+        # - the + 1 for _e is to shift the boundary by one chunk
+        # - this shifting is not needed if chunk_size divides e
+        _s, _e = s // chunk_size + p, e // chunk_size + p + (e % chunk_size
+                                                             > 0)
+
+        # adjust indices and offsets
+        chunk_indices[_s:_e] -= p
+        chunk_offsets[_s] = s % chunk_size
+
+    return chunk_indices, chunk_offsets
+
+
+# update_metadata computes metadata required by triton conv1d kernel
+# NOTE: argument type is removed for now to prevent circular dependency
+#       V0 metadata will eventually be obsolete and we can add back
+#       the correct type by then
```

Comment:
You can put the type in double quotation marks and then do the import only if `TYPE_CHECKING` is enabled. See for example:
https://github.com/vllm-project/vllm/blob/7ea5c73ad794290b1be12e41227b89941c05752c/vllm/v1/attention/backends/tree_attn.py#L21-L23
and
https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/tree_attn.py#L188